### ISSUE TITLE
twisted.web.client.Agent should use CertificateOptions

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6893]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)| @hynek reported|
|-|-|
|Trac ID|trac#6893|
|Type|enhancement|
|Created|2014-01-07 17:41:38Z|
|Branch|https://github.com/twisted/twisted/tree/agent-certificateoptions-6893|

All TLS development happens there now, ClientContextFactory has some [unfortunate defaults](https://gist.github.com/dreid/1c491ce3826d1c0541de).

<details><summary>Searchable metadata</summary>

```
trac-id__6893 6893
type__enhancement enhancement
reporter__hynek hynek
priority__normal normal
milestone__ 
branch__branches_agent_certificateoptions_6893 branches/agent-certificateoptions-6893
branch_author__hynek hynek
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1389116498000000 1389116498000000
changetime__1389214634000000 1389214634000000
version__None None
owner__hynek hynek
cc__hynek
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_1"></a>|@hynek commented|
|-|-|

(In [41142]) Branching to agent-certificateoptions-6893.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_2"></a>|@hynek commented|
|-|-|

I’ve changed Agent’s behavior and while adding tests, slightly refactored it’s old HTTPS ones in a way that seems more like what we strive for nowadays.

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|Alex commented|
|-|-|

This looks good to me. I don't understand whether the build bot issues are significant though, I'd like someone else to take a look at those.

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|Alex set owner to @hynek|
|-|-|

Went through the build bot stuff with glyph, appears to be all false positives. Please merge!

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_5"></a>|@hynek set status to `closed`|
|-|-|

(In [41150]) Merge agent-certificateoptions-6893.

Author: hynek
Reviewer: Alex
Fixes: [#6893](https://github.com/twisted/twisted/issues/6893)

Make twisted.web.client.Agent use twisted.internet.ssl.CertificateOptions as
context factory so it benefits from all further SSL enhancements that take
place there.
